### PR TITLE
rgw: implement ipv4 aws:SourceIp condition for bucket policy

### PR DESCRIPTION
**Pull request by joke-lee**
Signed-off-by: yuliyang <yuliyang@cmss.chinamobile.com>

### CODE REVIEW

**Comment by joke-lee** [comment]
hi, @cbodley ,@adamemerson ,do you mind to help review this ? thanks

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -603,7 +603,7 @@ rgw::IAM::Environment rgw_build_iam_environment(RGWRados* store,
     e.emplace("aws:SecureTransport", "true");
   }
 
-  i = m.find("HTTP_HOST");
+  i = m.find("REMOTE_ADDR");
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: the beast frontend doesn't set a `REMOTE_ADDR`, can you please add that? it looks pretty easy:
```patch
diff --git a/src/rgw/rgw_asio_client.cc b/src/rgw/rgw_asio_client.cc
index d59e7bb03b..39040a2607 100644
--- a/src/rgw/rgw_asio_client.cc
+++ b/src/rgw/rgw_asio_client.cc
@@ -75,6 +75,7 @@ void ClientIO::init_env(CephContext *cct)
   char port_buf[16];
   snprintf(port_buf, sizeof(port_buf), "%d", socket.local_endpoint().port());
   env.set("SERVER_PORT", port_buf);
+  env.set("REMOTE_ADDR", socket.remote_endpoint().address().to_string());
   // TODO: set SERVER_PORT_SECURE if using ssl
   // TODO: set REMOTE_USER if authenticated
 }
```

**Comment by qrGitHub** [comment]
Amusing! This one is almost the same with #19111 , but the initialization of "m.v6" can be optimized, could you please take a look at that one and make these changes?

**Comment by joke-lee** [comment]
@cbodley can this be merged ?

**Comment by yuriw** [comment]
[wip-yuri4-testing-2017-12-11-1549](https://trello.com/c/WwU6NRw5/209-wip-yuri4-testing-2017-12-11-1549)

**Pull request merged**